src/corvus.c

```diff
@@ -18,14 +18,33 @@
 #include "timer.h"
 #include "alloc.h"
 
+static pthread_spinlock_t signal_lock;
 static struct context *contexts;
 
-void sig_handler(int sig)
+void sigsegv_handler(int sig)
 {
-    int i;
+    if (pthread_spin_trylock(&signal_lock) == EBUSY) {
+        sigset_t mask;
+        sigemptyset(&mask);
+        pselect(0, NULL, NULL, NULL, NULL, &mask);
```

Comment:
If just return from sig handler the thread will continue running from the interrupted point from my understanding. If use pselect the thread will suspend and has no chance to run again.